Interactive Editor (#72)
bevyengine/rfcs: rfcs/72-editor.md

> # Feature Name: `editor-foundation`

Reviewer: First: I'm stoked that you're thinking about these things and there are a lot of good ideas here!

As a general comment, I think this RFC bites off more complexity than it should, in too many disparate areas, in not enough detail. So many of these things will be dictated by details that haven't yet been established. It feels like putting the cart before the horse. I think each of these categories deserves one or more RFCs, and each RFC should cover more details and more "design thoughts / justifications". I think the path forward is picking one of these areas that interests you, diving deep, and driving the scenario to completion. Another path forward is to omit any prescriptive details and we have an RFC whose purpose is just "agreeing on a small core list of things that need to be done for an editor MVP", but thats not really our design bottleneck right now. We need to sort out what specific scenarios look like.

1. **Default Project Structure**: this will largely be determined by the editor data model (which has not yet been established). If we roll with hot-reloading / dynamic linking, that will require a very different project structure than statically referencing. This deserves its own RFC, and should be covered only after contending with these details.
2. **Bevy Assets**:  This definitely deserves to be broken into one or more RFC design that (together) cover in detail the end-to-end "asset repository" scenario.  What will a repository "be" functionally (what protocol will it speak)? What will the manifest format look like and how will it be consumed (aka what _is_ an asset)? "Code assets" will need to play into the **Default Project Structure** scenario (either cover this after (1), or make sure it covers the range of possible paths).  Where will "downloaded code asset" code go? How will it be built? Do we distinguish between editor plugins, app plugins, and libraries? Is adding assets repeatable? Are users allowed to edit them? Are they stored adjacent to user assets (also plays into (1))? 
3. **Bevy Build**: Introducing a separate build command has massive implications for how people develop games in Bevy (lots of pros and lots of cons). This deserves way more than a couple of lines, both to describe what it does internally, why it does it that way, and plenty of rationale justifying its existence relative to `cargo`. One of the justifications for it is "bundling with all needed assets" (which I do agree with). But there are a ton of details there that have not yet been established. Will this be used for "dev" scenarios or just "final deployments"? If it is used for dev scenarios, will it try to avoid "re-doing" work (ex: copying expensive folders, rebuilding assets). Bevy Asset V2 doesn't yet have a CLI (but it will). How will this command relate to it? Do we need both apis? Why? One major justification for a separate build command is automatic cargo feature management. That deserves coverage and mentioning.
4. **Scene and Entity Manipulator**: We need something shaped roughly like this, but this deserves way more coverage than it is given before we can decide on what these systems are and how they relate to each other (ex: what is editable and when).

Author: > Another path forward is to omit any prescriptive details and we have an RFC whose purpose is just "agreeing on a small core list of things that need to be done for an editor MVP",

That's much closer to what I was going for.

> but thats not really our design bottleneck right now. We need to sort out what specific scenarios look like.

Ah, I see.

By scenarios, do you mean user stories?

Reviewer: > That's much closer to what I was going for.

I'm happy to roll with this path, but I'm relatively certain that as soon as we eliminate the "design details" we will trivially be in consensus. Ex: "yes we definitely want a bevy cli, an asset repo, a visual editor with a scene and entity inspector, a standard project structure, and plugin support". I think our time is probably better spent discussing what each of these things looks like in detail.

> By scenarios, do you mean user stories?

Kind of. From my perspective, "scenario" is a nebulous term that applies to either user stories or "feature areas". In this case, I used "scenario" to mean "specific implementable feature areas" like "bevy asset repos", "default project structure", "bevy build", "scene and entity manipulator", etc. "User stories" can cross cut through many of these "scenarios".

## Summary

This RFC outlines the foundation for Bevy's editor. It describes its basic
components and how they will interact with Bevy's engine. Following this RFC

> will result in a minimum viable product Bevy Editor.

Reviewer 2: ```suggestion
will result in a minimum viable product for the Bevy Editor.
```

This RFC does NOT cover the particular user interface (UI) design for the
editor. Instead, it provides the building blocks for a proper UI design, with a

> minimal usable interface to place these components in context.

Reviewer 2: This sentence isn't clear to me.

Author: How about:

Instead, it provides the minimum set of features required for the Bevy editor, and discusses basic ways these features compose. The full composition of these features, however, is left for a future RFC.

?

## Motivation

> Unlike most major game engines, Bevy lacks an interactive editor. This lack has

Reviewer 2: ```suggestion
Unlike most major game engines, Bevy lacks an interactive editor. This has
```

> three major downsides:

Reviewer 2: I think that "no good way to display visual elements like a schedule visualizer" is a key downside.

> 1. it increases the barrier to entry for potential new developers;

Reviewer 2: I don't think that an editor will meaningfully improve this.

Reviewer 3: I think the point here should be to help new developers visualize the running content, something that bevy_inspector_egui already does well for instance

Reviewer 4: I think the target here should be not the coders, but art designers.
Allowing them to preview, and in the future edit the graphics in an interactive way is the way.

> 2. this forces scene designers to learn to create scenes with code; and

Reviewer 2: I don't think this is the right way to frame it: it's not a learning curve problem. It's a "dramatically slower and more frustrating" problem.

> 3. developers must rebuild their project before seeing any changes they make.

Reviewer 2: The ";" make this really hard to read.

Author: oops, those used to be multi-part clauses, I can change them to be independent sentences.

While teams can mitigate the last two points by creating project specific
tools, this does require them to divert resources building those tools that they
could have spent building the game.

> Bevy's plugin-heavy design poses several challenges to making a traditional

Reviewer 2: I don't think the following paragraphs in this section are helpful. The points that they're trying to make aren't clear.

editor. Primarily, unlike other game engines, Bevy does limit developers access
to the core game loop, similar instead to many game frameworks. While an
interactive editor could be built on top of Bevy that takes away this freedom;
doing so violates the spirit of Bevy's plugin-heavy design, thus robbing it of
Bevy's expressive nature.

> **Bevy's editor must acknowledge the duel nature of Bevy as both a game engine

Reviewer 2: ```suggestion
**Bevy's editor must acknowledge the dual nature of Bevy as both a game engine
```

and a game framework.**

A secondary, smaller challenge, is that Bevy relies on the Cargo build system
for modularity. An interactive-editor that is core to Bevy must both respect
this design, and work with the limitations and abilities of Cargo.

## User-facing explanation

The easiest way to get started with Bevy's editor by downloading it with Cargo:

```

> $ cargo install bevy

Reviewer 5: Using `bevy` here is not possible because the `bevy` crate already exists.
```suggestion
$ cargo install bevy_cli
```
The actual binary being named `bevy` is good though :)

Author: Crates.io seems to let you have a binary and library crate have the same name. For example I just tested:

https://crates.io/crates/leiftestlibbinsplit

You can run `cargo install leiftestlibbinsplit` and run the executable and it runs just fine. And you can also make a new project and run `cargo add leiftestlibbinsplit`, and now you can `use leiftestlibbinsplit::*` in your project.

Am I missing something?

Reviewer 6: Sorry for the late reply and not being more specific.
It's possible to have a library and binary in the same crate but I don't think that's an option for Bevy. It would require syncing the release of a new version of Bevy's cli to a release of Bevy as a whole. We technically could do this but I figure this will cause us trouble sooner or later.

Reviewer 7: In any case I don't think having 2 things with the same name is good. This will create unnecessary confusion.

```

> (Note the use of `install`, not `add`!)

Reviewer 2: Explain the implication here.

Author: ```suggestion
*Note the use of `install`, not `add`!* This means Bevy's editor for your entire rust environment. In the future you may also be able to download it from external sources such as Steam.
```

This will add `bevy` to our path:

```
$ bevy
Bevy's Command Line Tool

Usage: bevy [OPTIONS] [COMMAND]

...

> ```

Reviewer 2: I think that the code block is imbalanced here.

This tool allows us to create new projects, manage assets, edit projects, and
even publish games. Note that `bevy` is not meant to replace `cargo`, but to
work alongside it.

> Also, just like Cargo, the Bevy version used in a project does not need to match

Reviewer 2: ```suggestion
Also, just like Cargo, the Bevy library version used in a project does not need to match
```

> the version of `bevy` you have installed. You can upgrade and downgrade them

Reviewer 2: ```suggestion
the version of the `bevy` CLI you have installed. You can upgrade and downgrade them
```

independently. Therefore, we recommend you always use the latest version of the
`bevy` tool.

Bevy's editor can be used graphically, or as a command line tool. This document
covers the command line usage, but running `bevy edit` in a project will open
the graphical view.

> **The `bevy` command line interface is not stable. Do not use it in build
> scripts!**

Reviewer 2: ```suggestion
**The `bevy` command line interface follows semver. Be careful when using it in build scripts!**
```

### Default Project Structure

The first entry point for the `bevy` command is `new` and `init`, for creating
new projects, meant to mimic the commands in cargo:

```
bevy new [options] <project-name>
```

As an example, let's create a new project, `flappers`:

```
$ bevy new flappers
```

This creates a new `flappers` folder in your current directory, using the
default template, similar to using `cargo new`. Unlike Cargo, however, a Bevy
project can start from any number of templates you can get from the Bevy asset
repository.

Next, let's take a look at the project structure:

```
flappers/
    README.md
    Cargo.toml

>     LICENSE-Apache2.txt

Reviewer 2: We should default to dual MIT + Apache.

Reviewer 8: I agree we shouldn't add license files in the template. I can see this turning into complaints when commits accidentally include unwanted licenses.

Reviewer 9: If you make the project creation a bit more interactive, then you could ask the user which kind of license, if any, should be added.
A bit like Rustup asks you which Rust version and modules should be installed.

Author: Mmm...perhaps this is a bit unclear on my part. What I imagined was that licensing is separate from the template, but still provided as a (changeable) default. The process for new project construction that I'm imagining is:

1. Fill out template.
2. Add chosen license
3. Add version control
4. Add CI template

This is because the license you choose to release your project under is (imo, usually) a separate concern from which template you use to start your project. And since the cross product of all licenses and templates makes sense, they should be seperate flags.

So something like:

```
bevy new -l mit -l gplv2 -t cardgame my-project
```

I do think we should by default add a license (unless the user explicitly passes something like `-l nolicense`), even if its an 'all rights reserved' statement.

Reviewer 5: It should definitely not add any license by default. That is unacceptable.

Licensing a Bevy project is niche enough that I'd leave it out of `bevy new` entirely.
`cargo new` doesn't provide this functionality either.

Author: Cargo doesn't, but these do:

1. Npm (Javascript) does.
2. Raco (Racket) does.
3. Leiningen (Clojure) does.
4. Bevy's existing template on github

Although notably these don't:
1. Sbt (Scala)
2. Pkg (Julia)
3. Godot

Another option is to create a LICENSE file and put text in it saying `<INSERT LICENSE>` as a default.

If we don't want to pick any particular license as a default, we can also just put something like `<Insert License Here>` as a default.

Reviewer 10: I would start without license and later on add a option to include license, because I still feel like the user should opt in for specifying the license, not the tool.

Also the vast majority of Bevy users will make closed source games, which normally is copyright, so no license by default fits the majority of use cases.

Godot, Unity and Unreal doesn't do that, and they should be our benchmark, not package managers, since Bevy is aiming to be a game editor.

    assets/
        ...
    src/
        main.rs
    tests/

>         ...

Author: Shrug, this seems like a thing that is decided by the template, not this tool. Although we can debate about if it should be in the default template, I could go either way.

    target/

>         ...

Reviewer 2: Yeah, I think we can cut this.

    editor/
        Cargo.toml
        src/
            main.rs
    .git/
        ...
    .github/

>         workflows/

Reviewer 9: We could make that part of one of the templates. Then user can get a default workflow for their CI, if a template for that exists.

>         workflows/

Author: See: https://github.com/bevyengine/rfcs/pull/72#discussion_r1390488808

            ...
```

The default template provides a lot more than a standard new cargo project. Much
of this is not strictly required, but will be wanted for the average project.
Because this is mostly convention, feel free to change the layout as needed. The
first three files: `RREADME.md`, `Cargo.toml`, and `LICENSE-Apache2.txt` are
standard meta-information that are in most crates. The next folder: `assets` is
the default location to place project non-code assets. The bulk of the source
code goes in the `src` directory, likewise tests go in `tests`. Project builds
go into the `target` directory.

> The `editor` directory provides a sub-crate for the Bevy editor. Associating

Reviewer 7: I don't understand this entire paragraph, sorry. What does `associating` mean here, how is that related with Bevy versions, and why is there a crate involved? I'm clueless. 🤔

Bevy's editor with the project allows you to have multiple projects using
different versions of Bevy in the same environment. The `bevy` command line tool

> delegates to this editor for any project specific actions. It also has the side

Reviewer 2: I think this last sentence would be clearer if split out into its own paragraph.

benefit of allowing you to more heavily tune the editor for your specific game
while still having access to the full suite of tools Bevy provides.

> The final directories are for version control management and game publishing.

Reviewer 2: We should just name these directly. Easier to follow, and very useful for beginners.

### Bevy Assets

The main repository for Bevy assets can be found on [the bevy webpage][asset-page]

Assets can be added to your project through the `bevy` tool:

```
bevy assets add <asset-name>
```

Likewise, you can remove the asset with:

```
bevy assets remove <asset-name>
```

In general, assets come in three varieties: *code*, *data*, and *template*. Code

> assets are little more than Cargo crates, many can even be installed with the

Reviewer 2: ```suggestion
assets are little more than Cargo crates, and many can even be installed with the
```

> `cargo` command.

Reviewer 2: ```suggestion
`cargo add` command.
```

Data assets are overlaid onto the existing project, and usually go into the
`assets` directory. Frequently these assets are large, should not be added to

> the repository directly, or should be loaded at build time from an external

Reviewer 2: I think we should consider just setting up git LFS by default?

source. Therefore, in addition to copying the assets into the project, data

> assets can also be references to an asset at an external location. **Note that

Reviewer 2: I don't understand this last sentence.

### Asset Repositories

In addition to the main repository for assets, projects can use additional
repositories for storing and retrieving assets. You can add a new repository with:

```
bevy repo add [repository-name] <repository-url>
```

This adds a repository to the current project. Optionally, a repository name can
be provided, which is used by `bevy assets`:

```
bevy assets add repository-name/asset-name
```

Despite their similarities asset repos are different from git remotes in one
crucial way: they are generally checked into the repository. In this sense, they
are more akin to git submodules or git large file storage than pure remotes.

Sometimes it makes sense to have a remote stored system wide, rather than with
any particular project. The `-g` flag allows you to add repositories globally.

```
bevy repo add -g [repository-name] <repository-url>
```

Alternatively, you can set the `BEVY_ASSET_REPO` environment variable to one or
more asset repositories.

### Publishing Tools

Bevy also helps you publish your games, either by building distributables, or
even having CI distribute builds each time its pushed.

Use `bevy build` to build your game:

```

> bevy build <target>

Reviewer 2: I think `target` should be optional: the default should build for Mac / Linux / Windows. And then allow configuration of default build targets.

Reviewer 4: What about `build.rs` and any sort of preprocessing/baking.
Would this be the command that does it?

As of right now it seems like a light wrapper around `cargo build`.
Worth noting, it should allow all of the arguments `cargo build` takes.
Possibly making default `cargo build` arguments a project configuration.

```

This will place your game, bundled with all needed assets, into your projects
`target` folder.

Alternatively, Bevy can publish a new build every time you push to a repository.
Right now this is only possible using GitHub and Itch.io, sign in with `bevy
login`:

```
$ bevy login github
Enter GitHub access token: ***
$ bevy login itch
Enter Itch.io access token: ***
```

**Note that these commands store the tokens in you system vault, not within the
repository.**

### Scene and Entity Manipulator

Bevy comes with an entity and scene manipulator. The easiest way to use these
tools is with the graphical editor, run `bevy editor` in the project directory.

The entity manipulator lists all of the entities in a given world, as well as
the components that are associated with those entities. You can add new entities
from this view, and attach new components to them. You can also edit the

> entities that were added using the manipulator.

Reviewer 2: Only the ones added using the manipulator?

You can view all entities in your world, not just ones created with the
manipulator. However, you are only guaranteed to be able to edit static

> entities, and entities created with the manipulator.

Reviewer 2: I don't understand the motivation here.

In addition to the entity manipulator, Bevy also provides developers with a
scene manipulator. This allows developers to place entities in a larger level,
and get a real-time view of how they interact with each other.

### Systems Manager

In addition to the manipulator, Bevy provides developers with a view of all the

> systems their games uses. In addition to creating new systems, developers can

Reviewer 2: I don't think "add new systems from the editor" makes sense for an MVP.

use this manager to find and open the code for existing systems. The systems
manager can also be viewed in the graphical editor, run `bevy editor` in the
project directory.

### Localization, Accessibility, Porting, and Plugins

Just like the rest of the Bevy engine, the editor can be extended using plugins.

> Some plugins are installed to the system, while others are added only to a

Reviewer 2: I don't understand this sentence.

project. Some possible plugin features are:

* Adding new sub-commands to the `bevy` CLI tool.
* Adding new views to the graphical editor.
* Adding context-specific renderings of component data.
* Localize both the editor and underlying game.
* Alter game behavior when porting to new platforms.

## Implementation strategy

The following is a list of features that are needed for the minimum viable
editor:

* Project/System separation
* Version dispatching
* CLI Interface
* Window Interface
* New project construction

> * Asset management

Reviewer 2: I don't think this is required for an MVP.

* Distributable packaging

> * Credential/secrets management

Reviewer 2: IMO publishing can be cut from an MVP.

> * Scene/Entity manipulator (with possible component editing dispatch).

Reviewer 2: I don't understand the parenthetical.

Reviewer 4: This should be what `bevy_inspector_egui` does currently, so exploration, filtering and editing component values.
By "dispatch" possibly suggesting a trait (object) with API for exposing the mutable values.
Does Bevy expose all the necessary interfaces for that or are they defined in the plugin crate?
Defining an API in Bevy would be a nice first step.

Reviewer 4: #62 probably related to this

> * Systems list

Reviewer 2: Wildly useful, but I'm not sure it's particularly required for an MVP. I don't think it's likely to impact architecture.

> * Plugin architecture

Reviewer 2: It's not clear what you mean here.

While it is possible to make a prototype with a subset of these features, each
feature implementation must (1) makes space for every other feature, and can (2)
assumes the other features exist.

The rest of this section will be a sketch out the rough implementation for each
of these features.

> ### Project/System separation

Reviewer 2: We should probably mention that "these crates live in the bevyengine/bevy repo" somewhere in this section.

> 1. Create two crates: `bevy_editor` and `bevy_editor_command`, the first will be

Reviewer 2: Bikeshed: `bevy_editor_command` doesn't clearly explain what the crate does to me.

   Call `cli` from that crate.
    1. This makes two `bevy` crates, a library one, which already exists, and an
       executable one, which will be what the user installs.
4. Add `bevy_editor_command` as a dependency for `bevy`, this can be done
   through the `bevy_internal` crate, or as an optional feature.
5. The `cli` function must do basic argument parsing, project templating, and
   system/credential management. It will dispatch everything else to the `bevy_editor` crate.

### Version dispatching

1. Any project using the `bevy` tool is expected to have an `editor` binary
   crate in their workspace, the template will fill this in automatically.
2. The dispatch can use `cargo`, as a simple example, `bevy editor` can become
   an alias for:
```
cargo run -p editor
```
3. For this reason, the `editor` crate must provide the main function, so the
   dispatch should ideally have graceful failing if it doesn't exist.
4. We are using a separate crate so that the end distributable does not need to
   rely on the editor, developers that _do_ want the deliverable to rely on the editor can call it themselves.

### CLI Interface

> 1. Simple command line argument parsing can be done with clap.
> 2. Simple error handling can be provided with anyhow.
> 3. There does not seem to be a standard pretty-output library for rust,
>    prettytable seems to be a start.
>    1. Options like crossterm might also fit the bill, but also seem like overkill.
> 4. The manipulator/live viewing components can delegate to the window interface.

Reviewer 2: ```suggestion
1. Simple command line argument parsing can be done with `clap`.
2. Simple error handling can be provided with `anyhow`.
3. There does not seem to be a standard pretty-output library for rust,
   `prettytable` seems to be a start.
   1. Options like `crossterm` might also fit the bill, but also seem like overkill.
4. The manipulator/live viewing components can delegate to the window interface.
```

### Window Interface

1. Extend from the base crate's world.
1. Ideally, everything should be do-able through the graphical interface, but
   some things can be pushed to later releases.
2. The `bevy_editor_command` window only needs to handle: (1)Credential Storage,
   (2) Template Downloads, and (3) Project Selection, everything else is handled
   by the a separate `bevy_editor` window.
3. The `bevy_editor_command` window must seamlessly transition to the
   `bevy_editor` window, as if opening a project.

> 4. The ui should be made using bevy components.

Reviewer 2: ```suggestion
4. The editor's UI should be made using `bevy_ui`.
```

### New project construction

1. Handled by `bevy_editor_command`.
2. Retrieve template from asset repo or local template storage.
3. Build template object mapping template languages to their result.
    1. For example foo.rs.tera should become foo.rs.
4. Fill in appropriate template fields from the user's environment.
5. Add CI workflows.
6. Add licenses.
7. Initialize as Git repo.

### Asset management

> 1. Template management can be done using terra.

Reviewer 2: ```suggestion
1. Template management can be done using `terra`.
```

> 2. Downloading static assets can be managed by reqwest.

Reviewer 2: ```suggestion
2. Downloading static assets can be managed by` reqwest`.
```

> 3. Downloading code assets can be done with cargo.

Reviewer 2: ```suggestion
3. Downloading code assets can be done with `cargo`.
```

> 4. Downloading template assets can be done with libgit2.

Reviewer 2: ```suggestion
4. Downloading template assets can be done with `libgit2`.
```

Reviewer 9: Maybe [gitoxide](https://github.com/Byron/gitoxide) is a better fit for this, one less wrapped C library. Makes builds more predictable and stable on different platforms.

> 5. Template assets should use `directories` to store system wide.

Reviewer 2: Not fully clear what this sentence means.

### Distributable Packaging

1. Bundle `cargo build` with built asset files.
2. Let developers create per-target custom build scripts.
3. Should ideally support WASM out of the box.

### Credential/secrets management

1. Should use the operating system's keychain/credential storage for storing
   secrets.
2. For now, only for uploading to itch.io in CI builds.

> ### Scene/Entity manipulator (with possible component editing dispatch).

Reviewer 2: Paranthetical is unclear.

> 1. Can use rfc 62 as its data model.

Reviewer 2: Definitely needs a link.

2. Should provide an editor for adding and removing components to entities.
3. Should also contain a basic editor for setting the initial values of an
   entity's components.
   1. As a bonus, plugins can allow for domain-specific editor for given
      component types.
4. The scene viewer should place all objects in a scene at their initial
   position.

> 5. A play button lets users run any given scene with the world's systems.

Reviewer 2: This is a major feature: I would like to see this split out into its own section.

Reviewer 4: Would this require defining how states and scenes are bound together?

### Systems list

1. Can use rfc 62 as its data model.
2. Can add an optional feature for Bevy's system to store metadata about the
   source location for systems, this will allow us to even find dynamic systems
   created at run-time.

### Plugin architecture

1. Most of the features listed above should be implemented as plugins.
2. This plugin-heavy approach allows future additions of i18n, l10n, and a11y.
3. a11y should use bevy_a11y
4. Internationalization can be done with either fluent or
   rust-i18n.
    1. If fluent, we should bring bevy_fluent or a similar project into the
       core project.

## Drawbacks

This is a lot of things for a single RFC. It would be possible to make a minimum
editor with fewer features, and would lead to an editor quicker.

However, the size of this RFC is merited for these reasons:

1. Most of these features are already required for making the editor usable.
2. Implementing most of these features can go faster when it assumes the other
   features are build along side it. For example, the metadata used for the
   entity manipulator can also be used for systems browser.

> 3. Making these features as plug-ins helps us make sure the plugin system is

Reviewer 2: Generally speaking, there's some ambiguity between "plugins in the sense of the `Plugin` trait" and "plugins in the sense of editor extensions". Unless these are intended to be strictly equivalent (which should be clarified), we should pick a different name.

   appropriately powerful.

> ## Rationale and alternatives

Reviewer 2: I think that each of these questions should get their own subsection with FAQ style headings. Currently, they're a bit too terse to properly understand.

> ## Rationale and alternatives

Reviewer 2: "Why should we use bevy_ui" is an important rationale
